### PR TITLE
[FIX] mail: well sanitize the mails coming from outlook desktop

### PR DESCRIPTION
**Pull request by mamatch**
### Steps
- Create a ticket with a customer ( it will send the email)
- Answer to the mail with Outlook Desktop( a multiple lines break)
- Look at the response in the discuss frame.

### Issue
In the frame there is much more lines break than in the original mail

Cause:
The format of outlook desktop mail before sanitizing looks like this
```
<div class="WordSection1">
<p class="MsoNormal">Test<o:p></o:p></p>
<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
<p class="MsoNormal">Two break lines<o:p></o:p></p>
<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
<div>
```
So when parsing it transforms the ``<o:p></o:p>`` in ``<p></p>`` which adds more lines break when displaying.

### Solution
Remove empty ``<o:...>`` and ``</o:...>`` which are specific to outlook desktop before sanitizing.

opw-3089550

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
seems overarching change, an empty p tag should be kept as far as I can tell. Removing it could change the style which is what you say is the cause of your change.

Looking at your content:

```html
<div class="WordSection1">
    <p class="MsoNormal">Test<p></p></p>
    <p class="MsoNormal"><p>&nbsp;</p></p>
    <p class="MsoNormal"><p>&nbsp;</p></p>
    <p class="MsoNormal">Two break lines<p></p></p>
    <p class="MsoNormal"><p>&nbsp;</p></p>
<div>
```
is fixed by chrome (because you can't have a P tag in a P tag) like this:

```html
<div class="WordSection1">
    <p class="MsoNormal">Test</p>
    <p></p>
    <p></p>
    <p class="MsoNormal"></p>
    <p>&nbsp;</p>
    <p></p>
    <p class="MsoNormal"></p>
    <p>&nbsp;</p>
    <p></p>
    <p class="MsoNormal">Two break lines</p>
    <p></p>
    <p></p>
    <p class="MsoNormal"></p>
    <p>&nbsp;</p>
    <p></p>
    <div></div>
</div>
```
so I agree that there is an issue, but I'm not sure how it can be solved, maybe doing as you did here but only for the outlook case (eg. removing empty `<o:p>...</o:p>` before the namespace is removed)

**Comment by mamatch** [comment]
I agree with you, I wonder if we have to change just for outlook specific mails or for all. I change it now

File: odoo/tools/mail.py [hunk 1]
```diff
@@ -190,6 +190,11 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
     src = src.replace(u'<%', misc.html_escape(u'<%'))
     src = src.replace(u'%>', misc.html_escape(u'%>'))
 
+    # On the specific case of Outlook desktop it adds unnecessary '<o:p></o:p>' tags which are parsed in '<p></p>'
+    # and shows additional lines break when displayed.
+    src = src.replace(u'<o:p>', u'')
+    src = src.replace(u'</o:p>', u'')
```
Comment (thiryr) [inline odoo/tools/mail.py:196]: Seems way too specific to put in the generic sanitizer. At least it should be a separate method but I don't think it should really be handled here at all.

<o:p> should not be converted into anything. Putting that tag in my browser directly just creates a 0-width element which shouldn't modify the appearance, afaict

I'll try replicating but seems odd. We'll also really want a test with this, whatever the solution ends up being.

**Review by thiryr — COMMENTED** [review]
Hey seems a bit louche, I think this probably deserves some more investigation.

Note that the XML in your example is invalid (`<div>` is opened twice, never closed)

File: odoo/tools/mail.py (hunk 1 above)
Comment (mamatch) [reply odoo/tools/mail.py:196]: For the unclosed <div> tags it's just because I crop that part that I though was helpful to understand the problem.

The <o:p> tag is specific to Microsoft Word to keep a certain format when transferring from word to HTML. 

When you say ", can you identify where that is done and maybe prevent this there instead?" you mean where the <o:p> is converted ?

Comment (thiryr) [reply odoo/tools/mail.py:196]: > cropped

Got it

> you mean where <o:p> is converted?

Yes, as far as I can tell if we just insert that into a page it should just remain unrendered, not transform into <p>
So if there's some point where we convert it, would probably be better to just not do that since the reason it's prefixed with with `o:` namespace is because it should only be rendered in outlook/word

Comment (nle-odoo) [reply odoo/tools/mail.py:196]: this happen in the cleaner.clean_html that calls lxml.html.fromstring, for example:

```py
> lxml.html.tostring(lxml.html.fromstring('<o:p>test</o:p>'))
b'<p>test</p>'
> lxml.html.tostring(lxml.html.fromstring('<o:p>test</o:p>', parser=lxml.html.HTMLParser(recover=False)))
*** lxml.etree.XMLSyntaxError: Tag o:p invalid, line 1, column 5
```
if we remove the [recover](https://lxml.de/4.1/api/lxml.etree.HTMLParser-class.html) option (`recover - try hard to parse through broken HTML (default: True)`) the code fails but I don't see an option to keep the "o:p" =/

Comment (thiryr) [reply odoo/tools/mail.py:196]: Thanks for having a look!
I see... if we don't use it, it crashes. If we use it, it breaks stuf... lxml is great

I could not find a way to get a clean parsed version of the document with lxml without first fixing the missing namespace (which could be numerous) so indeed we'll have to make do with a simpler approach.

I'd still like to do something more generic than just removing this specific problematic tag. I'll try having a look again tomorrow.

**Review by thiryr — COMMENTED** [review]
Seeing as a generic solution complexifies this sensitive code a bit too much, let's go with this for now and I'll keep a poc on the side if some other issue ever comes up. So you can keep the implementation as is :)

We'll also want a test with this. You can add one in `TestSanitizer` from the `test_mail` module

Cheers!

File: odoo/tools/mail.py [hunk 2]
```diff
@@ -190,6 +190,11 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
     src = src.replace(u'<%', misc.html_escape(u'<%'))
     src = src.replace(u'%>', misc.html_escape(u'%>'))
 
+    # On the specific case of Outlook desktop it adds unnecessary '<o:.*></o:.*>' tags which are parsed in '<p></p>'
+    # and break the sanitization of the mail body
+    o_tags = re.compile(r'</?o:.*?>')
```
[the remark this answers is not in the document]
Comment (mamatch) [reply odoo/tools/mail.py:195]: I think it's okay like that but I will look.I add the test now :)

File: odoo/addons/base/tests/test_mail.py [hunk 3]
```diff
@@ -86,6 +86,18 @@ def test_html(self):
...
+        cases = [
+            '<o:p> test 1 </o:p>',
+            '<o:tag1> test 1 </o:tag1>',
+            '<o:> test 1 </o:>',
+        ]
+
+        for content in cases:
+            html = html_sanitize(content)
+            self.assertNotIn('<o:', html)
+            self.assertNotIn('</o:', html)
```
Comment (thiryr) [inline odoo/addons/base/tests/test_mail.py:99]: Can you test the actual full result in a more realistic context? Having some parent/sibling elements, more similar to the example you give in the commit message. Just to ensure it doesn't mangle everything on accident, since it's a custom regex, not xml tree editing

File: odoo/tools/mail.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thiryr) [reply odoo/tools/mail.py]: :brain: Ah, of course. Good to know, thanks!

**Review by thiryr — COMMENTED** [review]
Commit message is not consistent with the code. We remove all `<o:...>` tags, not just the empty ones. I think that's fine, we should just update the commit message.

I'll also check what kind of content outlook can add to see if there's anything else we should be wary of.

File: odoo/tools/mail.py [hunk 4]
```diff
@@ -190,6 +190,10 @@ def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=Fals
     src = src.replace(u'<%', misc.html_escape(u'<%'))
     src = src.replace(u'%>', misc.html_escape(u'%>'))
 
+    # On the specific case of Outlook desktop it adds unnecessary '<o:.*></o:.*>' tags which are parsed in '<p></p>'
+    # and break the sanitization of the mail body
```
Comment (nle-odoo) [inline odoo/tools/mail.py:194]: the comment is a little unclear, I would say the tags are "replaced  by `<p></p>`" and "which may alter the appearance (eg. spacing) of the mail body"

also if possible (even if the other lines doesn't do it), it's better to cut the line at at most 100 characters

File: odoo/addons/base/tests/test_mail.py [hunk 5]
```diff
@@ -86,6 +86,30 @@ def test_html(self):
...
+        <br>
+        <b>Sent:</b> Monday, November 20, 2023 8:34 AM<br>
+        <b>To:</b> Eteil DJOUMATCHOUA &lt;dej@gmail.com&gt;<br>
+        <b>Subject:</b> test (#23)<o:p></o:p></p>
+        </div>
+        </div>
+        """
+
+        html = html_sanitize(case)
+        self.assertEqual(re.search(r'</?o:.*?>', html), None)
```
Comment (thiryr) [inline odoo/addons/base/tests/test_mail.py:111]: We're still not testing the actual output.
It could very well be an empty document.

I'm saying we can just copy the exact output. So we know at a glance that it's valid.

Comment (mamatch) [reply odoo/addons/base/tests/test_mail.py:111]: Oh okay, I didn't understand that like that. I add it :)

**Review by reth-odoo — APPROVED** [review]
Seems fine to me, a bit annoying that we don't preserve these for the purpose of emails but seeing as they're always empty nodes anyway it doesn't matter. And I guess the other namespaced nodes also get broken already, anyway

--------------------------------
### Fixed in newer versions of libxml2

Note that this issue is fixed in libxml2 version 2.11 and up here
https://gitlab.gnome.org/GNOME/libxml2/-/issues/508
https://gitlab.gnome.org/GNOME/libxml2/-/commit/d7d0bc6581e332f49c9ff628f548eced03c65189

So if other tags end up being problematic, it might be inconsistent across OS distributions

---------------------------------
### Documentation for additional tags

Also after testing, it seems outlook at least adds xml namespaces with references to the document (even though it's not valid, since the rest of the document is not necessarily xhtml), but the schema doesn't seem to be accessible so there's not any current definitive list of tags we could reference `https://schemas.microsoft.com/office/2004/12/omml`

from office outlook 2021
`xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:w="urn:schemas-microsoft-com:office:word"  xmlns:m=3D"http://schemas.microsoft.com/office/2004/12/omml" xmlns="http://www.w3.org/TR/REC-html40"`

Though I was able to track down `https://learn.microsoft.com/en-us/previous-versions/office/developer/office-2003/ee364377(v=office.11)` which is old reference documentation for office 2003

I don't think most of those tags will be any issue. `o:table`, possibly, but I didn't manage to insert one with modern outlook.

File: odoo/addons/base/tests/test_mail.py [hunk 6]
```diff
@@ -86,6 +86,46 @@ def test_html(self):
...
+<p class="MsoNormal">Here is a test mail<o:p></o:p></p>
+<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
+<p class="MsoNormal">With a break line<o:p></o:p></p>
+<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
+<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
+<p class="MsoNormal">Then two<o:p></o:p></p>
+<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
+<div>
+<div style="border:none;border-top:solid #E1E1E1 1.0pt;padding:3.0pt 0in 0in 0in">
+<p class="MsoNormal"><b>From:</b> Mitchell Admin &lt;dej@gmail.com&gt;
```
Comment (tde-banana-odoo) [inline odoo/addons/base/tests/test_mail.py:100]: Don't include "real" emails in demo / tests :) (change to dummy@example.com for example) (several emails to change)

File: odoo/addons/base/tests/test_mail.py [hunk 7]
```diff
@@ -86,6 +86,46 @@ def test_html(self):
...
+</div>"""
+
+        expected = """<div class="WordSection1">
+<p class="MsoNormal">Here is a test mail</p>
+<p class="MsoNormal"> </p>
+<p class="MsoNormal">With a break line</p>
+<p class="MsoNormal"> </p>
+<p class="MsoNormal"> </p>
+<p class="MsoNormal">Then two</p>
+<p class="MsoNormal"> </p>
```
Comment (tde-banana-odoo) [inline odoo/addons/base/tests/test_mail.py:115]: Isn't it the inverse (2 break lines, then one) ?

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, :socks: remarks otherwise LGTM :) and thanks @reth-odoo for followup

File: odoo/addons/base/tests/test_mail.py (hunk 7 above)
Comment (mamatch) [reply odoo/addons/base/tests/test_mail.py:115]: It's the break lines above XD

File: odoo/addons/base/tests/test_mail.py [hunk 8]
```diff
@@ -86,6 +86,46 @@ def test_html(self):
...
+<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
+<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
+<p class="MsoNormal">Then two<o:p></o:p></p>
+<p class="MsoNormal"><o:p>&nbsp;</o:p></p>
+<div>
+<div style="border:none;border-top:solid #E1E1E1 1.0pt;padding:3.0pt 0in 0in 0in">
+<p class="MsoNormal"><b>From:</b> Mitchell Admin &lt;dummy@example.com&gt;
+<br>
+<b>Sent:</b> Monday, November 20, 2023 8:34 AM<br>
+<b>To:</b> Eteil DJOUMATCHOUA &lt;dummy@example.com&gt;<br>
```
Comment (tde-banana-odoo) [inline odoo/addons/base/tests/test_mail.py:103]: Maybe also remove the name ^^ ?

**Review by tde-banana-odoo — APPROVED** [review]

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Comment by tde-banana-odoo** [comment]
@robodoo retry
